#  Segmenting and Clustering Neighborhoods in Toronto

# Web scraping for Toronto neighborhood and build a clean dataframe

In [64]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import requests
from pandas.io.json import json_normalize
import json

import requests

from bs4 import BeautifulSoup


import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

print('Libraries imported!')

Libraries imported!


# import the scraped data into dataframe

# Capstone project:  Segmenting and Clustering Neighborhoods in toronto City

## Description of the problem:

### In this capstone project I will be working with toronto city geo data and will be exploring the segmenting and Clustering function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. 

### Throughout the course we learnt various techniques to convert addresses into their equivalent latitude and longitude values. Also, will use the Foursquare API to explore neighborhoods in toronto City.

## Discussion of the background:

### In Module 3, we explored New York City and the city of Toronto and segmented and clustered their neighborhoods. Both cities are very diverse and are the financial capitals of their respective countries. 

### So, I would be interested in comparing the toronto neighborhoods of toronto cities and determine how similar or dissimilar they are.

### I will use the k-means clustering algorithm to complete this task. Finally, using the Folium library to visualize the neighborhoods in New York City and their emerging clusters. 


In [65]:
# open Wiki page with Beautiful Soup
data = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(data, 'html.parser')

## Description of the data:

### Web scraping for Toronto neighborhood and build a clean data frame:

#### Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe.

#### •	The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
#### •	Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
#### •	More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.
#### •	If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
#### •	Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
#### •	In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.
#### •	Note: There are different website scraping libraries and packages in Python. One of the most common packages is BeautifulSoup. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/


### •	The data we want is in a table, with 3 columns PostalCode, Borough and Neighborhood.
### •	The table contains a list of postal codes in Canada where the first letter is M. Postal codes beginning with M are located within the city of Toronto in the province of Ontario. Only the first three characters are listed, corresponding to the Forward Sortation Area.


In [66]:
postalCodeList = []
boroughList = []
neighborhoodList = []

for row in soup.find('table').find_all('tr'):
    cells = row.find_all('td')
    if(len(cells) > 0):
        postalCodeList.append(cells[0].text.rstrip('\n'))
        boroughList.append(cells[1].text.rstrip('\n'))
        neighborhoodList.append(cells[2].text.rstrip('\n')) # remove the new line char from neighborhood cell

### Put the scraped data into dataframe

## After scraping the html, we get 3 lists containing the PostalCodes, Boroughs and Neighborhoods in order as in Wiki page. We will put those 3 lists into a dataframe with corresponding columns name.

In [67]:
#The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
toronto_neighorhood = [('PostalCode', postalCodeList),
                      ('Borough', boroughList),
                      ('Neighborhood', neighborhoodList)]
toronto_df = pd.DataFrame.from_dict(dict(toronto_neighorhood))
toronto_df.head()

PostalCode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

In [68]:
#Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
toronto_df_dropna = toronto_df[toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
toronto_df_dropna.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

## Group neighborhoods by postal and borough

### There are some neighborhoods that belongs to the same postal code and borough (Ex: M5A Downtown Toronto has 2 neighborhoods Harbourfront and Regent Park). We will concat them into same row, seperated by a colon.

In [69]:
toronto_df_grouped = toronto_df_dropna.groupby(['PostalCode','Borough'], as_index=False).agg(lambda x: ','.join(x))
toronto_df_grouped.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

## Deal with Not assigned Neighborhood

### For M7A Queen's Park, there is no neighborhood assigned. We will replace the 'Not assigned' with the value of the corresponding Borough

In [70]:
#If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
toronto_df_grouped.loc[toronto_df_grouped['Neighborhood']=='Not assigned','Neighborhood']=toronto_df_grouped.loc[toronto_df_grouped['Neighborhood']=='Not assigned','Borough']
toronto_df_grouped.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [71]:
# use the .shape method to print the number of rows of your dataframe.
toronto_df_cleaned = toronto_df_grouped
toronto_df_cleaned.shape

(103, 3)

In [72]:
pwd

'C:\\Users\\sari2\\Downloads'

### How the data will be used to solve the problem:

#### Part 2: Getting coordinates and add to the Toronto DataFrame

#### Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

#### The problem with this Package is you must be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates.

#### geolocator = Nominatim(user_agent="tl-toronto-neigh")
#### postalList = toronto_df_cleaned['PostalCode'].values latList = [] longList = []
#### for post in postalList: location = None while(location is None): location = geolocator.geocode('{}, Toronto, Ontario'.format(post)) if(location != None): lat = location.latitude long = location.longitude print(post, lat, long) latList.append(lat) longList.append(long)
#### toronto_coors = [('Postal Code', postalList), ('Latitude', latList), ('Longitude', longList)] coors = pd.DataFrame.from_items(toronto_coors)

#### Note: Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data
#### The csv file format has:

#### 3 columns: Postal Code, Latitude and Longitude
#### 103 rows: corresponding to 103 postal codes in our toronto dataframe


In [73]:
coors = pd.read_csv('Geospatial_Coordinates.csv')

In [74]:
print(coors.shape)
coors.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### We will merge the two dataframes. To get the correct result independant on the order of data, we need to set indexes of two dataframes to its Postal Code columns

In [75]:
toronto_df_temp = toronto_df_cleaned.set_index('PostalCode')
coors_temp = coors.set_index('Postal Code')
toronto_df_coors = pd.concat([toronto_df_temp, coors_temp], axis=1, join='inner')

toronto_df_coors.index.name = 'PostalCode'
toronto_df_coors.reset_index(inplace=True)

print(toronto_df_coors.shape)
toronto_df_coors.head()

(103, 5)


PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [76]:
from geopy.geocoders import Nominatim

###  I will be working on the same data set with an idea to find the similarity between different places in the city of Toronto and segmented and clustered their neighborhoods and will be identifying how diverse and how similar or dissimilar they are.

# Part 3: Explore and cluster the neighborhoods in Toronto

In [77]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


In [78]:
#using Folium to create a Map of Toronto with Boroughs markers on top
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coors['Latitude'], toronto_df_coors['Longitude'], toronto_df_coors['PostalCode'], toronto_df_coors['Borough'], toronto_df_coors['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

In [79]:
#Reduce the number of Boroughs to explore To reduce the numbers of calls to FourSquare API, we will only explore boroughs that have Toronto in their names.
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(39, 5)


PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [80]:
#Using Folium to create a Map of Toronto with Boroughs markers on top
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

# Using FourSquare API to explore the Boroughs

In [81]:
CLIENT_ID = '5OSYV2RIXBB0OJCR10D3ARWKMTRDIJQXBR34JI5KAMFNL0VZ' # your Foursquare ID
CLIENT_SECRET = 'KHABM0SV4B3KEC3RI0POWXEB5BVFNVWBW0W0PMYPZVJWKWDZ' # your Foursquare Secret
VERSION = '20200712' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5OSYV2RIXBB0OJCR10D3ARWKMTRDIJQXBR34JI5KAMFNL0VZ
CLIENT_SECRET:KHABM0SV4B3KEC3RI0POWXEB5BVFNVWBW0W0PMYPZVJWKWDZ


In [82]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [83]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(1620, 9)


PostalCode       Borough                  Neighborhood  BoroughLatitude  \
0        M4E  East Toronto                   The Beaches        43.676357   
1        M4E  East Toronto                   The Beaches        43.676357   
2        M4E  East Toronto                   The Beaches        43.676357   
3        M4E  East Toronto                   The Beaches        43.676357   
4        M4K  East Toronto  The Danforth West, Riverdale        43.679557   

   BoroughLongitude                           VenueName  VenueLatitude  \
0        -79.293031                   Glen Manor Ravine      43.676821   
1        -79.293031  The Big Carrot Natural Food Market      43.678879   
2        -79.293031                 Grover Pub and Grub      43.679181   
3        -79.293031                       Upper Beaches      43.680563   
4        -79.352188                       MenEssentials      43.677820   

   VenueLongitude      VenueCategory  
0      -79.293942              Trail  
1      -79.297734  Health Food Store  
2      -79.297215                Pub  
3      -79.292869       Neighborhood  
4      -79.351265     Cosmetics Shop

In [84]:
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()

PostalCode  Borough           Neighborhood                                                                                              
M4E         East Toronto      The Beaches                                                                                                     4
M4K         East Toronto      The Danforth West, Riverdale                                                                                   42
M4L         East Toronto      India Bazaar, The Beaches West                                                                                 20
M4M         East Toronto      Studio District                                                                                                40
M4N         Central Toronto   Lawrence Park                                                                                                   4
M4P         Central Toronto   Davisville North                                                                                                7

In [85]:

len(venues_df['VenueCategory'].unique())

229

In [86]:
#analyzing venues in each area
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1620, 232)


PostalCode       Borough                 Neighborhoods  Airport  \
0        M4E  East Toronto                   The Beaches        0   
1        M4E  East Toronto                   The Beaches        0   
2        M4E  East Toronto                   The Beaches        0   
3        M4E  East Toronto                   The Beaches        0   
4        M4K  East Toronto  The Danforth West, Riverdale        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0     

In [87]:
#Get the frequency of occurance of each category in an area
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()

(39, 232)


PostalCode          Borough                   Neighborhoods  Airport  \
0        M4E     East Toronto                     The Beaches      0.0   
1        M4K     East Toronto    The Danforth West, Riverdale      0.0   
2        M4L     East Toronto  India Bazaar, The Beaches West      0.0   
3        M4M     East Toronto                 Studio District      0.0   
4        M4N  Central Toronto                   Lawrence Park      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0               0.0              0.00000           0.0       0.0          0.0   
1               0.0              0.02381           0.0       0.0          0.0   
2               0.0              0.00000           0.0       0.0          0.0   
3               0.0              0.05000           0.0       0.0          0.0   
4               0.0              0.00000           0.0       0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0         0.0                  0.0               0.0                 0.0   
1         0.0                  0.0               0.0                 0.0   
2         0.0                  0.0               0.0                 0.0   
3         0.0                  0.0               0.0                 0.0   
4         0.0                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop   Bakery   Bank    Bar  \
0            0.0        0.0         0.0         0.0  0.00000  0.000  0.000   
1            0.0        0.0         0.0         0.0  0.02381  0.000  0.000   
2            0.0        0.0         0.0         0.0  0.00000  0.000  0.000   
3            0.0        0.0         0.0         0.0  0.05000  0.025  0.025   
4            0.0        0.0         0.0         0.0  0.00000  0.000  0.000   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0               0.0                 0.0    0.0              0.0       0.0   
1               0.0                 0.0    0.0              0.0       0.0   
2               0.0                 0.0    0.0              0.0       0.0   
3               0.0                 0.0    0.0              0.0       0.0   
4               0.0                 0.0    0.0              0.0       0.0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0         0.0                 0.0     0.0            0.0    0.00000       0.0   
1         0.0                 0.0     0.0            0.0    0.02381       0.0   
2         0.0                 0.0     0.0            0.0    0.00000       0.0   
3         0.0                 0.0     0.0            0.0    0.02500       0.0   
4         0.0                 0.0     0.0            0.0    0.00000       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                   0.0             0.0  0.00000          0.00000       0.0   
1                   0.0             0.0  0.02381          0.02381       0.0   
2                   0.0             0.0  0.05000          0.00000       0.0   
3                   0.0             0.0  0.05000          0.00000       0.0   
4                   0.0             0.0  0.00000          0.00000       0.0   

   Burger Joint  Burrito Place  Bus Line  Butcher     Café  \
0           0.0           0.00      0.00      0.0  0.00000   
1           0.0           0.00      0.00      0.0  0.02381   
2           0.0           0.05      0.00      0.0  0.00000   
3           0.0           0.00      0.

In [88]:
#Get 10 most occurance venue types in each area
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
25        M5S  Downtown Toronto   
33        M6K      West Toronto   
3         M4M      East Toronto   
15        M5C  Downtown Toronto   
26        M5T  Downtown Toronto   
14        M5B  Downtown Toronto   
6         M4R   Central Toronto   
19        M5J  Downtown Toronto   
13        M5A  Downtown Toronto   
29        M5X  Downtown Toronto   
18        M5H  Downtown Toronto   
36        M6S      West Toronto   
28        M5W  Downtown Toronto   
16        M5E  Downtown Toronto   
37        M7A  Downtown Toronto   
20        M5K  Downtown Toronto   
11        M4X  Downtown Toronto   
9         M4V   Central Toronto   
21        M5L  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
5         M4P   Central Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
0         M4E      East Toronto   
34        M6P      West Toronto   
22        M5N   Central Toronto   
10        M4W  Downtown Toronto   
2         M4L      East Toronto   
4         M4N   Central Toronto   
23        M5P   Central Toronto   
8         M4T   Central Toronto   
31        M6H      West Toronto   
7         M4S   Central Toronto   
24        M5R   Central Toronto   
38        M7Y      East Toronto   

                                        Neighborhoods 1st Most Common Venue  \
27  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
32                           Little Portugal, Trinity                   Bar   
25                     University of Toronto, Harbord                  Café   
33       Brockton, Parkdale Village, Exhibition Place                  Café   
3                                     Studio District                  Café   
15                                     St. James Town                  Café   
26          Kensington Market, Chinatown, Grange Park                  Café   
14                           Garden District, Ryerson        Clothing Store   
6                  North Toronto West,  Lawrence Park        Clothing Store   
19  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
13                          Regent Park, Harbourfront           Coffee Shop   
29             First Canadian Place, Underground city           Coffee Shop   
18                           Richmond, Adelaide, King           Coffee Shop   
36                                 Runnymede, Swansea           Coffee Shop   
28                                     Stn A PO Boxes           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
37        Queen's Park, Ontario Provincial Government           Coffee Shop   
20           Toronto Dominion Centre, Design Exchange           Coffee Shop   
11                        St. James Town, Cabbagetown           Coffee Shop   
9   Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
21                     Commerce Court, Victoria Hotel           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
5                                    Davisville North      Department Store   
35                             Parkdale, Roncesvalles             Gift Shop   
1                        The Danforth West, Riverdale      Greek Restaurant   
30                                           Christie         Grocery Store   
0                                         The Beaches     Health Food Store   
34                      High Park, The Junction South    Mexican Restaurant   
22                                           Roselawn           Music Venue   
10                                           Rosedale                  Park   
2                      India Bazaar, The Beaches West                  Par

# Clustering areas

## Use KMeans algorigthm, try to cluster the toronto central areas into 4 clusters

In [89]:
kclusters = 4

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
10        M4W  Downtown Toronto   
4         M4N   Central Toronto   
23        M5P   Central Toronto   
8         M4T   Central Toronto   
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
25        M5S  Downtown Toronto   
33        M6K      West Toronto   
3         M4M      East Toronto   
15        M5C  Downtown Toronto   
26        M5T  Downtown Toronto   
14        M5B  Downtown Toronto   
6         M4R   Central Toronto   
19        M5J  Downtown Toronto   
13        M5A  Downtown Toronto   
29        M5X  Downtown Toronto   
18        M5H  Downtown Toronto   
36        M6S      West Toronto   
28        M5W  Downtown Toronto   
16        M5E  Downtown Toronto   
37        M7A  Downtown Toronto   
20        M5K  Downtown Toronto   
11        M4X  Downtown Toronto   
9         M4V   Central Toronto   
21        M5L  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
5         M4P   Central Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
34        M6P      West Toronto   
2         M4L      East Toronto   
31        M6H      West Toronto   
7         M4S   Central Toronto   
24        M5R   Central Toronto   
38        M7Y      East Toronto   
0         M4E      East Toronto   
22        M5N   Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
10                                           Rosedale  43.679563 -79.377529   
4                                       Lawrence Park  43.728020 -79.388790   
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
8                         Moore Park, Summerhill East  43.689574 -79.383160   
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
32                           Little Portugal, Trinity  43.647927 -79.419750   
25                     University of Toronto, Harbord  43.662696 -79.400049   
33       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
3                                     Studio District  43.659526 -79.340923   
15                                     St. James Town  43.651494 -79.375418   
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
14                           Garden District, Ryerson  43.657162 -79.378937   
6                  North Toronto West,  Lawrence Park  43.715383 -79.405678   
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
13                          Regent Park, Harbourfront  43.654260 -79.360636   
29             First Canadian Place, Underground city  43.648429 -79.382280   
18                           Richmond, Adelaide, King  43.650571 -79.384568   
36                                 Runnymede, Swansea  43.651571 -79.484450   
28                                     Stn A PO Boxes  43.646435 -79.374846   
16                                        Berczy Park  43.644771 -79.373306   
37        Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
11                        St. James Town, Cabbagetown  43.667967 -79.367675   
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
17                                 Central Bay Street  43.657952 -79.387383   
12                               Church and Wellesley  43.665860 -79.383160   
5                                    Davisville North  43.712751 -79.390197   
35                             Parkdale, Roncesvalles  43.648960 -79.456325   
1                        The Danforth West, Riverdale  43.679557 -79.352188   
30                                           Christie  43.669542 -79.422564   
34                      High Park, The Junction South  43.661608 -79.46476

In [90]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Cluster 0 (Red): Living area (with mostly park, trail, school, and some small businesses)
## Cluster 1 (Yellow): Roselawn - Central Toronto (nothing here except a garden)
## Cluster 2 (Purple): Business area (with lots of business venues)
## Cluster 3 (Blue) : Oldest neighboorhood with lot of nature park